evmap/src/aliasing.rs

```diff
@@ -0,0 +1,254 @@
+use std::cell::Cell;
+use std::marker::PhantomData;
+use std::mem::MaybeUninit;
+use std::ops::Deref;
+
+/// A `T` that is aliased across the two map copies.
+///
+/// You should be able to mostly ignore this type, as it can generally be treated exactly like a
+/// `&T`. However, there are some minor exceptions around forwarding traits -- since `Aliased` is a
+/// wrapper type around `T`, it cannot automatically forward traits it does not know about to `&T`.
+/// This means that if your `&T` implements, say, `Serialize` or some custom `Borrow<Q>`,
+/// `Aliased<T>` will not implement that same trait. You can work around this either by
+/// implementing your trait specifically for `Aliased<T>` where possible, or by manually
+/// dereferencing to get the `&T` before using the trait.
+#[repr(transparent)]
+pub struct Aliased<T> {
+    aliased: MaybeUninit<T>,
+
+    // We cannot implement Send just because T is Send since we're aliasing it.
+    _no_auto_send: PhantomData<*const T>,
+}
+
+impl<T> Aliased<T> {
+    /// Create an alias of the inner `T`.
+    ///
+    /// # Safety
+    ///
+    /// This method is safe because safe code can only get at `&T`, and cannot cause the `T` to be
+    /// dropped. Nevertheless, you should be careful to only invoke this method assuming that you
+    /// _do_ call [`drop_copies`] somewhere, as the `Aliased` is no longer safe to use after it is
+    /// truly dropped.
+    pub(crate) fn alias(&self) -> Self {
+        // safety:
+        //   We are aliasing T here, but it is okay because:
+        //    a) the T is behind a MaybeUninit, and so will cannot be accessed safely; and
+        //    b) we only expose _either_ &T while aliased, or &mut after the aliasing ends.
+        Aliased {
+            aliased: unsafe { std::ptr::read(&self.aliased) },
+            _no_auto_send: PhantomData,
+        }
+    }
+
+    /// Construct an aliased value around a `T`.
+    ///
+    /// Note that we do not implement `From<T>` because we do not want users to construct
+    /// `Aliased<T>`s on their own. If they did, they would almost certain end up with incorrect
+    /// drop behavior.
+    pub(crate) fn from(t: T) -> Self {
+        Self {
+            aliased: MaybeUninit::new(t),
+            _no_auto_send: PhantomData,
+        }
+    }
+}
+
+// Aliased gives &T across threads if shared or sent across thread boundaries.
+// Aliased gives &mut T across threads (for drop) if sent across thread boundaries.
+// This implies that we are only Send if T is Send+Sync, and Sync if T is Sync.
+//
+// Note that these bounds are stricter than what the compiler would auto-generate for the type.
+unsafe impl<T> Send for Aliased<T> where T: Send + Sync {}
+unsafe impl<T> Sync for Aliased<T> where T: Sync {}
+
+// XXX: Is this a problem if people start nesting evmaps?
+// I feel like the answer is yes.
+thread_local! {
+    static DROP_FOR_REAL: Cell<bool> = Cell::new(false);
+}
```

Comment:
Instead of storing a boolean flag, if all the values are stored in a contiguous block, you could store a pointer range, and check if the current value is in that range to check if they need to be dropped.

Reply:
Hmm, how do you envision we'd go about storing all the values in a contiguous block?

Reply:
I wasn't sure if that was the case, so if it's not then I'm not sure how to resolve nested evmaps.

Reply:
Ah, no, value storage is very much non-continguous at the moment. The values are stored entirely independently.

Reply:
While this is a very convenient and cool way of doing it, it might be a bit too much magic and risky. Considering one needs to remember to call `drop_copies()` and keep the guard alive anyway, it might make more sense (and not be more error-prone) to just make dropping of values explicit in `absorb_second`.

Reply:
You mean the thread-local more generally? The challenge is that there often isn't a way to make the dropping explicit in `absorb_second` without adding _significant_ complexity. For example, imagine you're doing a `retain` operation on the map — how do you ensure that the values are _not_ dropped in `absorb_first`, but _are_ dropped in `absorb_second`?

Reply:
I don't think it would be that big of an issue, to be honest. Specifically `retain` would be pretty simple because it's already wrapping its predicate, and something like `clear` would need to use a draining iterator calling it. The drop behaviour is definitely more convenient, but it leaves us with the thread-local problem - I'll talk about that more in response to my other comment.

Reply:
I think I may have a non-thread-local way to do this: https://github.com/rust-lang/unsafe-code-guidelines/issues/35#issuecomment-735491663

Reply:
If I'm understanding that correctly, that's pretty close to the first option I proposed in my other comment, yes? Just with a second wrapper type rather than casting to `T` directly - and if it's not correct to cast `Aliased<T>` to `T` directly despite it being `repr(transparent)`, a second wrapper type is probably the way to go.

Reply:
Yup, and it's close to what evmap does today, but trying to do it in a way that might be safe. Unfortunately it _does_ require some additional trickery. The trait bounds required... aren't great.